### ISSUE TITLE
Accumulating GPU memory when creating new meshes in env.reset

### ISSUE THREAD

**Issue by BlGene** [haosulab/ManiSkill#965]
Hello,

Thanks for the great work with ManiSkill.

I have a question: I'm resetting an environment very often (200+ times), which is leading to memory being accumulated on the GPU. During my reset, I'm loading new meshes in the  `_load_scene` function. I can "fix" the memory accumulation by deleting and recreating the env. I've tried deleting the objects during the reset, but this does not help, how can I fix this problem?

Below are my snippets from the env.reset and the builder function.

```
    def reset(self, seed: Union[None, int, list[int]] = None, options: Union[None, dict] = None):
        if self.object_dataset == "clevr":
            pass
        elif self.object_dataset == "ycb" or self.object_dataset == "objaverse":
            if self.objects is not None:
                for object in self.objects:
                    object.remove_from_scene()
                    del object
        else:
            raise ValueError
        #set_trace()
        return super().reset(seed=seed, options=options)
 ```

```
    if add_collision:
        collision_file = glb_path
        # builder.add_nonconvex_collision_from_file # This does not support density
        builder.add_multiple_convex_collisions_from_file(
            filename=collision_file,
            scale=[scale] * 3,
            # material=physical_material,
            density=density,
            pose=obj_pose,
        )

    if add_visual:
        visual_file = glb_path
        builder.add_visual_from_file(
            filename=visual_file, scale=[scale] * 3, pose=obj_pose
        )
```

**Comment by StoneT2000** [comment]
An environment reset shouldn't accumulate gpu memory, it should stay around the same.

Are you resetting normally or resetting with reconfigure True? Or did you set reconfigure freq to 1?

What do your load scene and initialize episode functions look like?

**Comment by BlGene** [comment]
```obs, _ = env.reset(seed=args.seed[0], options=dict(reconfigure=True))```
I reset manually after each episode.

```python
def _load_scene_objaverse(self, num_objects: int=2):

    from mani_skill.examples.objaverse_handler import SpokDatasetBuilderFast,SpokDatasetBuilder, get_spok_builder
    if self.spok_dataset is None:
        self.spok_dataset = SpokDatasetBuilderFast()

    uuids = self.spok_dataset.sample_uuids(num_objects, with_replacement=False)


    for uuid in uuids:
        obj_builder = get_spok_builder(self.scene, uuid, add_collision=True, add_visual=True,
                                       spok_dataset=self.spok_dataset)
        model_name = f"{uuid}"
        shape_name = f"{uuid}"
        shape_name = self.spok_dataset.get_object_name(uuid) 


        self.objects.append(obj_builder.build(name=f"{model_name}"))
        self.objects_descr.append(dict(size="", color="", shape=shape_name))


    unique_vals, counts = np.unique(uuids, return_counts=True)
    count_dict = dict(zip(unique_vals, counts))
    self.objects_unique = [count_dict[num] == 1 for num in uuids]

def _load_scene(self, options: dict):
    if self.scene_dataset == "Table":
        self.table_scene = TableSceneBuilder(env=self, robot_init_qpos_noise=self.robot_init_qpos_noise)
        self.table_scene.build()
    elif self.scene_dataset == "ProcTHOR":
.... 
    else:
        raise ValueError

    min_objects = 2
    max_objects = 5
    num_objects = int(randomization.uniform(float(min_objects), float(max_objects+1), size=(1,)))
    
    self.objects = []
    self.objects_descr = []
    if self.object_dataset == "clevr":
        self._load_scene_clevr(num_objects)
    elif self.object_dataset == "ycb":
        self._load_scene_ycb(num_objects)
    elif self.object_dataset == "objaverse":
        self._load_scene_objaverse(num_objects)
    else:
        raise ValueError
    
    self.cubeA = self.objects[0]  
    self.cubeB = self.objects[1]
```

... In the `intervention` function, I call object.pose.get_p(), sometimes without .clone().detach().

```python
def _initialize_episode(self, env_idx: torch.Tensor, options: dict):
    with torch.device(self.device):
        b = len(env_idx)
        self.table_scene.initialize(env_idx)

        xyz = torch.zeros((b, 3))
        if self.scene_dataset == "ProcTHOR":
...
        else:
            xyz[:, 2] = 0.0 #0.02
            xy = torch.rand((b, 2)) * 0.2 - 0.1 # rand is uniform [0,1], so shift to [-0.1, 0.1]

        sampler = randomization.UniformPlacementSampler(bounds=self.object_region[:2].T, batch_size=b)
        
        for shape, descr in zip(self.objects, self.objects_descr):
            # Get radius of object
            radius = np.linalg.norm(get_actor_obb(shape).primitive.extents) / 2
            if self.scene_dataset == "ProcTHOR":
...
            else:
                shape_xy = xy + sampler.sample(radius, max_trials=100, verbose=False)
                xyz[:, :2] = shape_xy

            qs = randomization.random_quaternions(
                b,
                lock_x=True,
                lock_y=True,
                lock_z=False,
            )
            shape.set_pose(Pose.create_from_pq(p=xyz.clone(), q=qs))
        
        obj_start, obj_end = intervention(self, pretend=True)
        self.set_goal_pose(obj_end)
        self.obj_start = obj_start
        self.obj_end = obj_end
        from mani_skill.examples.clevr_env_solver import get_grasp_pose_and_obb    
        grasp_pose, _ = get_grasp_pose_and_obb(self)
        self.grasp_pose = Pose.create_from_pq(p=grasp_pose.get_p(), q=grasp_pose.get_q())

        are_visible = False
        if self.RANDOMIZE_VIEW:
            self.initalize_render_camera_random()
            for i in range(self.RESAMPLE_CAMERA_IF_OBJS_UNSEEN):
                camera = self.scene.human_render_cameras['render_camera'].camera 
                are_visible = self.check_objects_visible(obj_start, obj_end, camera)
                if are_visible:
                    break
                self.initalize_render_camera_random()
                self._setup_sensors(options)
            if not are_visible:
                print("Warning: could not sample visible camera position.")
            
        else:
            self.initalize_render_camera_fixed()
```

**Comment by StoneT2000** [comment]
Hm I don't see anything that is problematic yet.

Best bet for yourself right now is to manually comment out the code and add them line by line until there is no more gpu memory leaks

**Comment by fffffq00** [comment]
@StoneT2000  
I'm using the version `3.0.0b20` of `ManiSkill`.
I have also encountered a similar problem. My environment needs to be initialized frequently due to the high probability of failure. I found that as the epoch increases, my GPU memory usage increases until it overflows. I didn't use reconfigure because I was doing a partial reset. To alleviate this issue, I manually used `gc.collect()` to release GPU memory at the end of each rollout, and found that the memory was significantly reduced. 
I'm not sure what the specific reason is. My task and initialization method are similar to the `pick_cube` environment, only with more frequent resets. Can you provide some help?

**Comment by StoneT2000** [comment]
@fffffq00 when running the original PickCube task do you have any memory overflow issues?

What do you mean exactly by frequent resets? Do you have an example reproducible script?

**Comment by fffffq00** [comment]
@StoneT2000   Thanks for your reply.
I found no memory overflow when running the original `PickCube` task.
Frequent resets mean that I set the `fail` condition to be too large force between robot joints or collision with the table, so in the early stage of RL, the environment will be frequently reset. Because the project is relatively cumbersome, I have not yet organized the sample code.
However, I tested some situations again today, and I found that when I manually simplified almost all collision bodies into capsules and boxes, the GPU memory usage would be reduced, and memory would rarely accumulate. (Before this, in order to make the collision body similar to the visual, I split the visual mesh into multiple sub-colliders as the collision body during simulation)
Can you analyze the reason through this situation?

**Comment by StoneT2000** [comment]
Do you have frequent resets with reconfiguration? Or no reconfiguration?


And what are the more complex collision shapes like? Any picture?

Based on the description this is either caused by your code or physx has a bug

**Comment by fffffq00** [comment]
@StoneT2000  Hi, 
I didn't use reconfiguration because I needed a partial reset.
I'm using a shadowhand. The middle part of the finger in this picture is the collision body I used before. 
![Image](https://github.com/user-attachments/assets/36b7b1c0-c596-4dd1-b2d7-28df06d4f805)
Now I have simplified it to a box. After simplification, my code runs normally, but I will check my code further.
Thank you.

**Comment by StoneT2000** [comment]
Do you mind sharing your full code for the shadowhand?

A number of people want to use it in our framework so I might try and help model it for people.

**Comment by fffffq00** [comment]
@StoneT2000 Glad to hear that, I organized some code and put it together in a repository [shadowhand-maniskill](https://github.com/fffffq00/shadowhand-maniskill) where you can find the original configuration and my modified configuration of shadowhand. Hope it can help others.

**Comment by StoneT2000** [comment]
I will move this discussion to a new appropriately named issue for shadow hand modelling in #1073

**Comment by LiuYiKai6** [comment]
> [@fffffq00](https://github.com/fffffq00) when running the original PickCube task do you have any memory overflow issues?
> 
> What do you mean exactly by frequent resets? Do you have an example reproducible script?

I found the cpu memory accumulated slowly when running the original PickCube task. About 0.2GB for 1M steps. It this normal?

**Comment by fffffq00** [comment]
> I found the cpu memory accumulated slowly when running the original PickCube task. About 0.2GB for 1M steps. It this normal?

@LiuYiKai6  Hi, do you use CPU to train your single environment? I haven't tried this, but when I run this task on a GPU, it initially accumulates some GPU memory, and then it doesn't accumulate anymore.
